### PR TITLE
🌱 introduce a new version (v3) of the logicalcluster package

### PR DESCRIPTION
**Pull request by p0lyn0mial**
## Summary
This PR introduces a new version (v3) of the logicalcluster package.

The new version splits the old type into a new `Name` type (holds a value that uniquely identifies a logical cluster).

And a new `Path` type which represents a colon separated list of words describing a path in a logical cluster hierarchy,
like a file path in a file-system.

```
// For instance, in the following hierarchy:
//
// root/                    (62208dab)
// ├── accounting           (c8a942c5)
// │   └── us-west          (33bab531)
// │       └── invoices     (f5865fce)
// └── management           (e7e08986)
//     └── us-west-invoices (f5865fce)
//
// the following would all be valid paths:
//
//   - root:accounting:us-west:invoices
//   - 62208dab:accounting:us-west:invoices
//   - c8a942c5:us-west:invoices
//   - 33bab531:invoices
//   - f5865fce
//   - root:management:us-west-invoices
//   - 62208dab:management:us-west-invoices
//   - e7e08986:us-west-invoices
```

## Related issue(s)

Fixes #

### CODE REVIEW

File: path.go [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+	}
+	n.value = s
+	return nil
+}
+
+func (n Path) HasPrefix(other Path) bool {
+	return strings.HasPrefix(n.value, other.value)
+}
+
+const lclusterNameFmt string = "[a-z0-9]([a-z0-9-]{0,61}[a-z0-9])?"
```
Comment (p0lyn0mial) [inline path.go:125]: shouldn't we do validation on the `Name` type as well?

File: path.go [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+	"regexp"
+	"strings"
+)
+
+const (
+
+	// ClusterHeader set to "<lcluster>" on a request is an alternative to accessing the
+	// cluster via /clusters/<lcluster>. With that the <lcluster> can be access via normal kube-like
+	// /api and /apis endpoints.
+	ClusterHeader = "X-Kubernetes-Cluster"
```
Comment (p0lyn0mial) [inline path.go:31]: not sure what it was used for

File: path.go [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+func (n Path) Split() (parent Path, name string) {
+	i := strings.LastIndex(n.value, separator)
+	if i < 0 {
+		return Path{}, n.value
+	}
+	return Path{n.value[:i]}, n.value[i+1:]
+}
+
+// Base returns the last component of the logical cluster name.
+func (n Path) Base() string {
```
Comment (p0lyn0mial) [inline path.go:95]: does it become `func (p Path) Name() Name` ?

File: path.go [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+	return n, n.IsValid()
+}
+
+// Empty returns true if the logical cluster value is unset.
+func (n Path) Empty() bool {
+	return n.value == ""
+}
+
+// Path returns a path segment for the logical cluster to access its API.
+func (n Path) Path() string {
```
[the remark this answers is not in the document]
Comment (p0lyn0mial) [reply path.go:67]: feels like this should be part of a client library

File: path.go (hunk 2 above)
Comment (sttts) [reply path.go:31]: maybe we got rid of it?

Comment (sttts) [reply path.go:31]: @stevekuznetsov

File: path.go [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+
+// Path represents a logical cluster. A logical cluster is
+//  1. a (part of) etcd prefix to store objects in that cluster
+//  2. a (part of) a http path which serves a Kubernetes-cluster-like API with
+//     discovery, OpenAPI and the actual API groups.
+//  3. a value in metadata.clusterName in objects from cross-workspace list/watches,
+//     which is used to identify the logical cluster.
+//
+// A logical cluster is a colon separated list of words. In other words, it is
+// like a path, but with colons instead of slashes.
```
Comment (sttts) [inline path.go:44]:
```suggestion
// Path represents a colon separated list of words descring a path in a logical cluster hierarchy,
// like a file path in a file-system.
```

File: path.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (sttts) [reply path.go:67]: rename it to RequestPath?

File: path.go (hunk 3 above)
Comment (sttts) [reply path.go:95]: no. Normally, this does not even return a logical cluster name. Only the first element in a path is usually (not necessarily, there might be exceptions) a cluster name.

File: path.go (hunk 1 above)
Comment (sttts) [reply path.go:125]: yes, in the same way with a `IsValid` func.

File: path.go (hunk 3 above)
Comment (p0lyn0mial) [reply path.go:95]: does it mean we need a new `func (p Path) Name() Name` method that simply takes the whole path as a `Name` i.e. `Name{path.value}`

File: name.go [hunk 6]
```diff
@@ -54,31 +27,19 @@ var (
...
-// A valid logical cluster returns true on IsValid().
-func NewValidated(value string) (Name, bool) {
-	n := Name{value}
-	return n, n.IsValid()
-}
-
-// Empty returns true if the logical cluster value is unset.
-func (n Name) Empty() bool {
-	return n.value == ""
+// Name represent a cluster name.
```
Comment (sttts) [inline name.go:30]:
```
// Name represent a cluster name. A cluster name
// 1. can be used to access a cluster via `/cluster/<cluster-name>`
// 2. is part of an etcd key path
// 3. is used to uniquely reference a logical cluster. There is at most one cluster name for a logical cluster, but many logical cluster string can point to the same cluster name.
```

File: path.go (hunk 3 above)
Comment (sttts) [reply path.go:95]: `func (p Path) Name() (Name, bool)`

File: path.go (hunk 2 above)
Comment (stevekuznetsov) [reply path.go:31]: Remove: https://github.com/kcp-dev/kcp/pull/2448

File: name.go (hunk 6 above)
Comment (p0lyn0mial) [reply name.go]: done

File: path.go (hunk 5 above)
Comment (p0lyn0mial) [reply path.go]: done

File: path.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (p0lyn0mial) [reply path.go]: done

File: path.go (hunk 1 above)
Comment (p0lyn0mial) [reply path.go:127]: done

**Comment by p0lyn0mial** [comment]
PTAL, it is `WIP` on fixing/adding unit tests.

File: name.go [hunk 7]
```diff
@@ -16,69 +16,46 @@ limitations under the License.
...
+
+	clusterNameRegExp = regexp.MustCompile(clusterNameString)
 )
 
-// New returns a Name from a string.
-func New(value string) Name {
-	return Name{value}
-}
+const (
+	clusterNameString string = "^[a-z0-9]([a-z0-9-]{0,61}[a-z0-9])?"
```
Comment (stevekuznetsov) [inline name.go:35]: What's the utility of this indirection?

File: name.go [hunk 8]
```diff
@@ -16,69 +16,46 @@ limitations under the License.
...
+)
 
-// NewValidated returns a Name from a string and whether it is a valid logical cluster.
-// A valid logical cluster returns true on IsValid().
-func NewValidated(value string) (Name, bool) {
-	n := Name{value}
-	return n, n.IsValid()
+// Name represent a cluster name. A cluster name
```
Comment (stevekuznetsov) [inline name.go:38]:
```suggestion
// Name represents a cluster name, which:
```

File: name.go [hunk 9]
```diff
@@ -16,69 +16,46 @@ limitations under the License.
...
+// 1. can be used to access a cluster via `/cluster/<cluster-name>`
+// 2. is part of an etcd key path
+// 3. is used to uniquely reference a logical cluster. There is at most one cluster name for a logical cluster, but many logical cluster string can point to the same cluster name.
```
Comment (stevekuznetsov) [inline name.go:41]:
```suggestion
// 3. is used to uniquely reference a logical cluster. There is at most one cluster name for a logical cluster, but many logical cluster identifiers can alias one cluster name.
```

File: name.go [hunk 10]
```diff
@@ -16,69 +16,46 @@ limitations under the License.
...
+type Name struct {
+	value string
 }
 
-// Empty returns true if the logical cluster value is unset.
-func (n Name) Empty() bool {
-	return n.value == ""
+// NewName returns a Name from a string.
```
Comment (stevekuznetsov) [inline name.go:46]:
```suggestion
// NewName returns a Name.
```

File: name.go [hunk 11]
```diff
@@ -16,69 +16,46 @@ limitations under the License.
...
-	return path.Join("/clusters", n.value)
+// Path returns a Path (a logical cluster) out of the Name
+func (n Name) Path() Path {
+	return New(n.value)
 }
 
-// String returns the string representation of the logical cluster name.
-func (n Name) String() string {
-	return n.value
+// IsValid returns true if the name is a Wildcard starts with a lower-case letter and contains only lower-case letters, digits and hyphens.
```
Comment (stevekuznetsov) [inline name.go:56]:
```suggestion
// IsValid returns true if the name is a Wildcard or starts with a lower-case letter and contains only lower-case letters, digits and hyphens.
```

File: path.go (hunk 2 above)
Comment (stevekuznetsov) [reply path.go:31]: Still need to remove

File: path.go [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+
+	separator = ":"
+)
+
+// Path represents a colon separated list of words describing a path in a logical cluster hierarchy, like a file path in a file-system.
+type Path struct {
+	value string
+}
+
+// New returns a Path from a string.
```
Comment (stevekuznetsov) [inline path.go:41]:
```suggestion
// New returns a Path.
```

File: path.go [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+func New(value string) Path {
+	return Path{value}
+}
+
+// NewValidated returns a Path from a string and whether it is a valid logical cluster.
```
Comment (stevekuznetsov) [inline path.go:46]:
```suggestion
// NewValidated returns a Path and whether it is valid.
```

File: path.go [hunk 14]
```diff
@@ -0,0 +1,135 @@
...
+// A valid logical cluster returns true on IsValid().
```
Comment (stevekuznetsov) [inline path.go:47]:
```suggestion
```

File: path.go [hunk 15]
```diff
@@ -0,0 +1,135 @@
...
+func NewValidated(value string) (Path, bool) {
+	n := Path{value}
+	return n, n.IsValid()
+}
+
+// Empty returns true if the logical cluster value is unset.
```
Comment (stevekuznetsov) [inline path.go:53]:
```suggestion
// Empty returns true if the path is unset.
```

**Review by stevekuznetsov — COMMENTED** [review]
This needs a large package-level GoDoc blob with clear examples. Kind of hard to follow right now.

**Comment by stevekuznetsov** [comment]
My suggestion for clarity:

```go
// Identifier uniquely identifies a logical cluster and remains stable for the lifetime of the logical cluster.
// For instance, a logical cluster may have the identifier 62208dab
type Identifier string


// Name is a human-readable identifier for a logical cluster, relative to the logical cluster's parent.
// For instance, in a parent logical cluster 62208dab, a child logical cluster may
// be referred to by the Name "staging".
type Name string

// Path uniquely identifies a logical cluster in a hierarchy by listing the identifiers for parent logical clusters from the root.
// Each path segment identifies a logical cluster, and may be either an Identifier (for which no further parents may be listed),
// or a Name (for which the parent must be listed).
// For instance, in the following hierarchy the Names are listed first, with Identifiers in parentheses:
// root/ (62208dab)
// └── organization (c8a942c5)
//     └── team (33bab531)
//         └── project (f5865fce)
//
// The following would all be valid Paths for the `project` logical cluster:
// root:organization:team:project
// 62208dab:organization:team:project
// c8a942c5:team:project
// 33bab531:project
// f5865fce
type Path []PathSegment

// PathSegment identifies a logical cluster.
type PathSegment interface {
    Identifier | Name
}
```

File: path.go [hunk 16]
```diff
@@ -0,0 +1,135 @@
...
+	separator = ":"
+)
+
+// Path represents a colon separated list of words describing a path in a logical cluster hierarchy, like a file path in a file-system.
+type Path struct {
+	value string
+}
+
+// New returns a Path from a string.
+func New(value string) Path {
```
Comment (sttts) [inline path.go:42]: should we turn that into `NewPath`?

File: path.go [hunk 17]
```diff
@@ -0,0 +1,135 @@
...
+
+// Join joins a parent logical cluster name and a name component.
+func (n Path) Join(name string) Path {
+	if n.value == "" {
+		return Path{name}
+	}
+	return Path{n.value + separator + name}
+}
+
+func (n Path) MarshalJSON() ([]byte, error) {
```
Comment (sttts) [inline path.go:110]: we need the same for Name, no?

File: path.go [hunk 18]
```diff
@@ -0,0 +1,135 @@
...
+}
+
+const lclusterNameFmt string = "[a-z0-9]([a-z0-9-]{0,61}[a-z0-9])?"
+
+var lclusterRegExp = regexp.MustCompile("^" + lclusterNameFmt + "(:" + lclusterNameFmt + ")*$")
+
+// IsValid returns true if the name is a Wildcard or a colon separated list of words where each word
+// starts with a lower-case letter and contains only lower-case letters, digits and hyphens.
+func (n Path) IsValid() bool {
+	return lclusterRegExp.MatchString(n.value)
```
Comment (sttts) [inline path.go:134]: || wildcard

**Comment by stevekuznetsov** [comment]
@p0lyn0mial LMK when you've incorporated https://github.com/kcp-dev/logicalcluster/pull/27#issuecomment-1335525671 (`Name` can be unexported, right?) and I will re-review

File: name.go [hunk 19]
```diff
@@ -16,73 +16,45 @@ limitations under the License.
...
-	TODO = None
+const (
+	clusterNameString string = "^[a-z0-9][a-z0-9-]{0,61}[a-z0-9]$"
 )
 
-// New returns a Name from a string.
-func New(value string) Name {
-	return Name{value}
-}
+// Name represents a cluster name, which:
```
Comment (stevekuznetsov) [inline name.go:29]: This comment introduces a couple of concepts: `name`, `Name`, `cluster`, `logical cluster`, `identifiers` - we _need_ a package-level comment block that lays out the context and defines things.

File: path.go [hunk 20]
```diff
@@ -0,0 +1,154 @@
...
+		return Path{name}
+	}
+	return Path{n.value + separator + name}
+}
+
+func (n Path) MarshalJSON() ([]byte, error) {
+	return json.Marshal(&n.value)
+}
+
+func (n *Path) UnmarshalJSON(data []byte) error {
```
Comment (p0lyn0mial) [inline path.go:133]: can we remove this method ?

File: path.go [hunk 21]
```diff
@@ -0,0 +1,154 @@
...
+	var s string
+	if err := json.Unmarshal(data, &s); err != nil {
+		return err
+	}
+	n.value = s
+	return nil
+}
+
+func (n Path) HasPrefix(other Path) bool {
```
Comment (p0lyn0mial) [inline path.go:142]: can we remove this method ? i haven't found any traces of it in kcp repo

File: name.go [hunk 22]
```diff
@@ -16,73 +16,45 @@ limitations under the License.
...
+//     there is at most one cluster name for a logical cluster,
+//     but many logical cluster identifiers can alias one cluster name.
+type Name string
 
-// NewValidated returns a Name from a string and whether it is a valid logical cluster.
-// A valid logical cluster returns true on IsValid().
-func NewValidated(value string) (Name, bool) {
-	n := Name{value}
-	return n, n.IsValid()
+// Path returns a Path (a logical cluster) out of the Name.
```
Comment (stevekuznetsov) [inline name.go:37]: This comment is the sort of thing I was referring to with not being specific. We use `Name`, `Path` and `logical cluster` all in one tiny sentence, and we imply that `Path` is `logical cluster`, but it's not ... Let's please be more clear.

File: name.go [hunk 23]
```diff
@@ -16,73 +16,45 @@ limitations under the License.
...
-	return n.value == ""
+// String returns the string representation of the cluster name.
+func (n Name) String() string {
+	return string(n)
 }
 
-// Path returns a path segment for the logical cluster to access its API.
-func (n Name) Path() string {
-	return path.Join("/clusters", n.value)
+// IsValid returns true if the name starts with a lower-case letter and contains only lower-case letters, digits and hyphens.
```
Comment (stevekuznetsov) [inline name.go:47]: Comments should have the implementation details here - what's the utility to the user of knowing that? The method determines if the cluster is valid.

File: name.go [hunk 24]
```diff
@@ -16,73 +16,45 @@ limitations under the License.
...
+func (n Name) IsValid() bool {
+	return clusterNameRegExp.MatchString(string(n))
 }
 
-// String returns the string representation of the logical cluster name.
-func (n Name) String() string {
-	return n.value
+// Empty returns true if the cluster name is unset.
```
Comment (stevekuznetsov) [inline name.go:52]: Similarly, let's focus on what it means to be empty and what the user should expect from that. Commenting that this one-liner method does what it looks like it does it not useful to the person reading the comment.

**Comment by p0lyn0mial** [comment]
> This comment introduces a couple of concepts: name, Name, cluster, logical cluster, identifiers - we need a package-level comment block that lays out the context and defines things.

I would like to avoid a package-level comment block that defines the terms, instead I simplified the documenation a bit. Please see if it now makes more sense.

File: name.go [hunk 25]
```diff
@@ -16,140 +16,57 @@ limitations under the License.
...
-	TODO = None
+const (
+	clusterNameString string = "^[a-z0-9][a-z0-9-]{0,61}[a-z0-9]$"
 )
 
-// New returns a Name from a string.
-func New(value string) Name {
-	return Name{value}
-}
+// Name holds a value that uniquely identifies a logical cluster, which:
```
Comment (p0lyn0mial) [inline name.go:29]: It seems that the only term that could be documented is logical cluster, do we want to do that?

File: path.go [hunk 26]
```diff
@@ -0,0 +1,142 @@
...
+// like a file path in a file-system.
+//
+// For instance, in the following hierarchy:
+//
+// root/                    (62208dab)
+// ├── accounting           (c8a942c5)
+// │   └── us-west          (33bab531)
+// │       └── invoices     (f5865fce)
+// └── management           (e7e08986)
+//
```
Comment (sttts) [inline path.go:44]: no new line

File: path.go [hunk 27]
```diff
@@ -0,0 +1,142 @@
...
+//	└── us-west-invoices    (f5865fce)
+//
```
Comment (sttts) [inline path.go:46]:
```suggestion
// For example in the case the logical cluster hierarchy is implemented via workspaces,
// the first word is a logical cluster name, and the following words are workspace names.
// For instance, consider the following example paths:
```

File: name.go [hunk 28]
```diff
@@ -16,140 +16,57 @@ limitations under the License.
...
+//  1. can be used to access a cluster via `/clusters/$value`
+//  2. is part of an etcd key path
+type Name string
 
-// NewValidated returns a Name from a string and whether it is a valid logical cluster.
-// A valid logical cluster returns true on IsValid().
-func NewValidated(value string) (Name, bool) {
-	n := Name{value}
-	return n, n.IsValid()
+// Path creates a new Path object from the stored value.
```
Comment (sttts) [inline name.go:34]: from the name.

Comment (sttts) [reply name.go:34]: no strong opinion

File: name_test.go [hunk 29]
```diff
@@ -16,87 +16,39 @@ limitations under the License.
...
-
-func TestIsValidCluster(t *testing.T) {
+func TestIsValidName(t *testing.T) {
 	tests := []struct {
 		name  string
 		valid bool
 	}{
 		{"", false},
-		{"*", true},
+		{"*", false},
```
[the remark this answers is not in the document]
Comment (sttts) [reply name_test.go:27]: right

**Comment by p0lyn0mial** [comment]
/hold

I need a tag from @stevekuznetsov

File: name.go (hunk 25 above)
Comment (stevekuznetsov) [reply name.go:29]: Likely some value in a package-level documentation blob, yes. The way you have this comment written now, though, is 100x better!

File: path.go [hunk 30]
```diff
@@ -0,0 +1,163 @@
...
+)
+
+const (
+	separator = ":"
+)
+
+// Path represents a colon separated list of words describing a path in a logical cluster hierarchy,
+// like a file path in a file-system.
+//
+// For example in the case the logical cluster hierarchy is implemented via workspaces,
```
Comment (stevekuznetsov) [inline path.go:38]: What's a workspace? :)

File: name.go (hunk 25 above)
Comment (stevekuznetsov) [reply name.go:29]: If you noticed my straw-man GitHub comment used the same example data to explain Name, Identifier and Path - having one concrete example and running through how every part of this library applies would be great. That can either be split between the GoDoc for Name and Path, or a package-level thing, whatever.

Comment (stevekuznetsov) [reply name.go:29]: Notably, an example of a name would be good.

File: path.go [hunk 31]
```diff
@@ -0,0 +1,163 @@
...
+// the following would all be valid paths for the `project` logical cluster:
+//   - root:accounting:us-west:invoices
+//   - 62208dab:accounting:us-west:invoices
+//   - c8a942c5:us-west:invoices
+//   - 33bab531:invoices
+//   - f5865fce
+//   - root:management:us-west-invoices
+//   - 62208dab:management:us-west-invoices
+//   - e7e08986:us-west-invoices
+type Path struct {
```
Comment (stevekuznetsov) [inline path.go:59]: Why do we want this to be a struct? I thought @sttts preferred `type Path string` for fewer allocs?

File: name.go [hunk 32]
```diff
@@ -16,140 +16,57 @@ limitations under the License.
...
-
-const separator = ":"
+import "regexp"
 
 var (
-	// Wildcard is the name indicating cross-workspace requests.
-	Wildcard = New("*")
-
-	// None is the name indicating a cluster-unaware context.
-	None = New("")
```
Comment (stevekuznetsov) [inline name.go:51]: We lost None and TODO - they are useful for automated refactor tools (TODO) and will be critical when one controller code-base is to be able to run inside of kcp and also against k8s (None)

Comment (p0lyn0mial) [reply name.go:51]: ok, i'll bring it back.

File: path.go (hunk 31 above)
Comment (p0lyn0mial) [reply path.go:59]: i'll try with `type Path string` I think it should work.
unless you remember why the old type was a struct ?

File: path.go (hunk 30 above)
Comment (p0lyn0mial) [reply path.go:38]: i'll bring back the old comment

File: name.go (hunk 25 above)
Comment (p0lyn0mial) [reply name.go:29]: okay, i'll try to document a logical cluster and give an example for a name. Thanks.

File: path.go (hunk 30 above)
Comment (sttts) [reply path.go:38]: This is an example logical cluster hierarchy. There is no predefined hierarchy in this package. Hence, if we want to show these paths, we must formulate it as "for instance".

File: path.go (hunk 31 above)
Comment (sttts) [reply path.go:59]: We cannot change this here in the PR. Please leave the struct for now.

File: path.go (hunk 30 above)
Comment (p0lyn0mial) [reply path.go:38]: the only term will be a `logical cluster`, the rest (Name and Path) will be structured around it. I want to keep it very simple and short.

File: path.go [hunk 33]
```diff
@@ -0,0 +1,167 @@
...
+//
+// root/                    (62208dab)
+// ├── accounting           (c8a942c5)
+// │   └── us-west          (33bab531)
+// │       └── invoices     (f5865fce)
+// └── management           (e7e08986)
+//
+//	└── us-west-invoices    (f5865fce)
+//
+// the following would all be valid paths for the `project` logical cluster:
```
Comment (sttts) [inline path.go:54]: why 'project' ?

Comment (p0lyn0mial) [reply path.go:54]: can change to `the following would all be valid paths:`

**Comment by p0lyn0mial** [comment]
/hold cancel

@stevekuznetsov I've addressed the last batch of your comments. Let me know if there is anything you would like to improve/change and I will open a new PR.
